src/createSignRequest.js

```diff
@@ -0,0 +1,59 @@
+'use strict'
+const constructSignatureString = require('./constructSignatureString')
+const constructDigestString = require('./constructDigestString')
+
+const createSignRequest = (creationOptions) => {
+  const defaultOptions = {
+    constructSignatureString,
+    constructDigestString,
+    signedHeaders: ['(request-target)', '(created)'],
+    expiryOffset: 300,
+    digestEncoding: 'base64',
+    digestAlgorithm: 'sha-512',
+    ...creationOptions
+  }
+  return (req, options) => {
+    const combinedOptions = {
+      ...defaultOptions,
+      ...options
+    }
+
+    const { headers = {} } = req
+
+    const {
+      constructSignatureString,
+      constructDigestString,
+      keyId,
+      algorithmName,
+      signedHeaders,
+      expiryOffset,
+      digestAlgorithm
+    } = combinedOptions
+
+    if (!keyId || typeof keyId !== 'string') {
+      throw new Error('keyId in options object is missing or not a string')
+    }
+    if (!algorithmName || typeof algorithmName !== 'string') {
+      throw new Error('algorithmName in options object is missing or not a string')
+    }
+    if (!signedHeaders || !Array.isArray(signedHeaders) || signedHeaders.length < 1) {
+      throw new Error('signedHeaders must be an array with at least one value')
+    }
+
+    combinedOptions.created = Math.floor(Date.now() / 1000)
+
+    if (signedHeaders.includes('(expires)')) {
+      combinedOptions.expires = combinedOptions.created + expiryOffset
+    }
+
+    if (signedHeaders.includes('digest')) {
+      headers.digest = `${digestAlgorithm}=${constructDigestString(req, combinedOptions)}`
+    }
+
+    headers.Signature = `keyId="${keyId}", algorithm="${algorithmName}", headers="${signedHeaders.join(' ')}", signature="${constructSignatureString(req, combinedOptions)}", created=${combinedOptions.created}${combinedOptions.expires ? `, expires=${combinedOptions.expires}` : ''}`
```

Comment:
This assumes `headers` will be a plain object, and not a `Headers` class. I have first hand experience of attempting to add headers to a `GraphQLRequest` in this manor, and having ApolloServer ignore them (needed to use `headers.set`). We could add some sort of `instanceof` logic, or change this function to return signed headers and leave it up to the user to add them to their request. Thoughts or suggestions @eadmundo @jkirkpatrick24 ?

Reply:
At first, I was passing in a Request object (which included an instance of a Headers class and a URL class) but I changed it to a simple object based on feedback from @jkirkpatrick24 . The rationale is that the input and output of the signing method is as simple as possible and we can allow the developer using the method to format the output in whatever way they need.

Reply:
But the output is their, now mutated, request object. If we want to keep things simple, we should probably just return the signature header, and not modify the request at all.

Reply:
To clarify, I'm not saying we should just return the header or that we shouldn't mutate the request. I just think that if we are going to mutate the request, we should respect the headers' type.

Maybe we provide a `setHeader` option that defaults to:
```js
const defaultSetter = (carrier, key, value) => carrier[key] = value
```
Then if someone wants to use `Headers.set`:
```js
const customSetter = (carrier, key, value) => carrier.set(key, value)
```

Reply:
This is another very good point. Is this an issue and feature addition or do we need to add Header class compatibility before the initial release?